src/function/matrix/eigs/complex.js

```diff
@@ -0,0 +1,316 @@
+                    const rowMulRadix = big ? rowNorm.mul(radix) : rowNorm*radix
+
+                    if (!big) {
+                        while (c < rowDivRadix) {
+                            c *= radixSq
+                            f *= radix
+                        }
+                        while (c > rowMulRadix) {
+                            c /= radixSq
+                            f /= radix
+                        }
+                    } else {
+                        while (c.lessThan(rowDivRadix)) {
+                            c = c.mul(radixSq)
+                            f = f.mul(radix)
+                        }
+                        while (c.greaterThan(rowMulRadix)) {
+                            c = c.div(radixSq)
+                            f = f.div(radix)
+                        }
+                    }
+
+
+                    // check whether balancing is needed
+                    const condition = !big
+                        ? (c + rowNorm)/f < 0.95*(colNorm + rowNorm)
+                        : c.add(rowNorm).div(f).lessThan(colNorm.add(rowNorm).mul(0.95))
+
+
+                    // apply balancing similarity transformation
+                    if (condition)
+                    {
+                        // okay bro we should definitely loop once again
+                        last = false
+
+                        let g = big ? bigOne.div(f) : 1/f
+                        Rdiag[i] = big ? R[i].mul(f) : Rdiag[i]*f
+
+                        for (let j = 0; j < N; j++)
+                        {
+                            if (i === j) continue;
+                            arr[i][j] = multiply(arr[i][j], f)
+                            arr[j][i] = multiply(arr[j][i], g)
+                        }
+                    }
+
+                }
+            }
+        }
+
+        // return the diagonal transformation matrix
+        return diag(Rdiag)
+    }
+
+
+    /**
+     * @param {number[][]} arr
+     * @param {number} N
+     * @param {number} prec
+     * @param {'number'|'BigNumber'|'Complex'} type
+     * @param {number[][]} R
+     */
+    function reduceToHessenberg(arr, N, prec, type, R)
+    {
+        const big  = type === 'BigNumber'
+        const cplx = type === 'Complex'
+        const num = !big && !cplx
+
+        const bigZero = bignumber(0)
+
+        for (let i = 0; i < N-2; i++)
+        {
+            // Find the largest subdiag element in the i-th col
+
+            let maxIndex = 0
+            let max = big ? bigZero : 0
+
+            for (let j = i + 1; j < N; j++)
+            {
+                let el = arr[j][i]
+                if (big ? abs(max).lessThan(abs(el)) : abs(max) < abs(el)) {
+                    max = el
+                    maxIndex = j
+                }
+            }
+
+            // This col is pivoted, no need to do anything
+            if (big ? abs(max).lessThan(prec) : abs(max) < prec)
```

Comment:
Ideally, inside a function you should not have to worry about the type of a value (number/BigNumber), if you use functions which can handle multiple types. I see quite some `big ? ... : ...` and `num ? ... : ...` cases here. Would it be possible to replace these with functions like `smaller(max, prec)` instead so you don't have to worry here and can focus on the actual logic?